Add 'crystals competition' task for 'multithreading' module

danikk812 opened the pull request.

danikk812 committed:
Add 'crystals competition' task for 'multithreading' module

danikk812 committed:
Apply fixes to javacore.multithreading.crystals.competition.Magician

Fix logProgress() output to one-line per magician type

dark11ght commented on src/com/dmdev/javacore/multithreading/crystals/ApplicationRunner.java:
```diff
@@ -0,0 +1,26 @@
+public class ApplicationRunner {
+
+    public static void main(String[] args) {
+        CrystalStorage crystalWarehouse = new CrystalStorage();
```
так кто по итогу выиграл? 

и как так получается что у тебя соревнования закончиои не начавшись?))

<img width="566" alt="Screenshot 2024-03-18 at 19 57 21" src="https://github.com/danikk812/dmdev-java-core/assets/32550069/19d8ca1c-bec1-46ea-87e5-40382428d5fd">

dark11ght commented on src/com/dmdev/javacore/multithreading/crystals/ApplicationRunner.java:
```diff
@@ -0,0 +1,26 @@
+
+        Magician fireMagician = new Magician(MagicianType.FIRE, crystalWarehouse, redCrystalsCollected, whiteCrystalsCollected);
+        Magician airMagician = new Magician(MagicianType.AIR, crystalWarehouse, redCrystalsCollected, whiteCrystalsCollected);
+
```
на сколько я понял ты просто джоинишь потоки. и тут вопрос. а точно утебя гонка? или просто потоки независимо друг от друга собирают кристалы и потом просто один поток ждет другой пока не выполнится условие?

danikk812 committed:
Apply fixes to javacore.multithreading.crystals.competition classes

Fix Magician & CrystalStorage classes logic to meet the "magicians compete for shared crystal recourses" condition
